pkg/apis/crd/v1alpha2/types.go

```diff
@@ -221,3 +226,51 @@ type EgressList struct {
 
 	Items []Egress `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExternalIPPool defines one or multiple IP sets that can be used in external network. For instance, the IPs can be
+// allocated to the Egress resources as the Egress IPs.
+type ExternalIPPool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the ExternalIPPool.
+	Spec ExternalIPPoolSpec `json:"spec"`
+}
+
+type ExternalIPPoolSpec struct {
+	// The IP sets of this IP pool, e.g. 10.10.0.0/24, 10.10.10.2-10.10.10.20, 10.10.10.30-10.10.10.30.
+	IPSets []IPSet `json:"ipSets"`
+	// The Nodes that the external IPs can be assigned to. If empty, it means all Nodes.
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector"`
+}
+
+// IPSet is a set of contiguous IP addresses, represented by a CIDR or an IP range.
+type IPSet struct {
```

Comment:
I am hesitating whether we should call it IPSubnet, and later we can add fields like Gateway.

And if we go this way, CIDR and IPRange can co-exist, and it means a range of IPs from the subnet defined by the CIDR.

How other solutions define IPPool for allocating Pod IPs?

Reply:
My [original proposal ](https://github.com/antrea-io/antrea/issues/2128#issuecomment-843240955)was CIDR and IPRange co-existing and CIDR is mandatory and IPRange is optional but must be in the CIDR if set. I thought you meant providing either CIDR or IP Range is enought in external IP case but maybe I misunderstood.
CNI IPAM's [ippool](https://github.com/containernetworking/plugins/blob/72f2a1ffd4f6aa2355a67eb6746f74baca878fa6/plugins/ipam/host-local/backend/allocator/config.go#L65) :
```
type Range struct {
	RangeStart net.IP      `json:"rangeStart,omitempty"` // The first ip, inclusive
	RangeEnd   net.IP      `json:"rangeEnd,omitempty"`   // The last ip, inclusive
	Subnet     types.IPNet `json:"subnet"`
	Gateway    net.IP      `json:"gateway,omitempty"`
}
```
MetalLB's ingress IP pool:
```
addresses:
- 192.168.10.0/24
or it could be
addresses:
- 192.168.1.240-192.168.1.250
```

I think if this struct needs to be used for Pod IP case and Gateway field is potentially needed, CIDR should be mandatory. If IPRange is not set, all IPs in the CIDR are available by default (except .0, .1, .255), and user can specify IPRange to supply more or less IPs.

Reply:
Ok. What you think about:
```
type IPRange struct {
    CIDR string
    Start string
    End string
}
```
Is it strange, for external IP pool, only one of CIDR and Start/End is needed; but for Pod IP, CIDR is a must?

Reply:
Even for external IP pool, is it acceptable to make CIDR a must, which implictly specifies the start and the end as the regular range, e.g. x.2-x.254 for /24 case? I feel unifying the usage and validation is easier to explain and understand. When user allocates a range of IPs for external traffic, it's likely allocated from a subnet or a subset of it too?

Reply:
But: 1) for external IP pool, if you already have a range, why you still need CIDR? 2) for both cases, if often I just define a whole CIDR, why bother to provide IP range too?

Reply:
1) It's not required for external IP case technically, I was trying to simplify the validation and documentation of the API as most IPPool definition requires CIDR first then range optionally. I don't have strong opinion about it.
2) For Pod IP, IP range is often required when not all IPs in the subnet are supposed to be used by the Pool's consumers. That's why it's present in the spec of CNI IPAM? https://github.com/containernetworking/cni/blob/7ba00c594ca847f52630752cbd754ab3553daad7/CONVENTIONS.md#well-known-capabilities

Reply:
If the API requires only one of CIDR and Start+End, what if user specify both of them? should we reject it or validate whether they match? which one to use if they don't match?

Reply:
If we can use Range is both are set, and also check Range falls into the CIDR; but also allows only one is set?

Reply:
Updated, could you take another look?